### PR TITLE
Change string rendering to be chunked.

### PR DESCRIPTION
**Pull request by NTaylorMullen**
- Roslyn currently has an issue where too large of strings result in out of memory exceptions at compile time. To combat this I broke down literal strings into 1k pieces each resulting in their own `WriteLiteral`/`WriteLiteralTo` calls.
- Added tests to validate large string rendering.

#614

### CODE REVIEW

File: src/Microsoft.AspNet.Razor/CodeGenerators/Visitors/CSharpCodeVisitor.cs [hunk 1]
```diff
@@ -16,6 +15,7 @@ public class CSharpCodeVisitor : CodeVisitor<CSharpCodeWriter>
         private const string ItemParameterName = "item";
         private const string ValueWriterName = "__razor_attribute_value_writer";
         private const string TemplateWriterName = "__razor_template_writer";
+        private static readonly int MaxStringLiteralLength = 499; // Max 1k string size despite x86/x64
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/CodeGenerators/Visitors/CSharpCodeVisitor.cs]: Decided on 1k. Definitely open to other suggestions, this in all reality is just one of the many magic numbers that fit in the small object heap.

File: src/Microsoft.AspNet.Razor/CodeGenerators/CSharpCodeWriter.cs [hunk 2]
```diff
@@ -470,23 +470,6 @@ private void WriteCStyleStringLiteral(string literal)
                         Write(literal[i].ToString());
                         break;
                 }
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/CodeGenerators/CSharpCodeWriter.cs]: If we want we could always render C# strings always as `@"hello world"` or `"\r\nhello\tworld"` but thought I'd reduce the amount of change by just nuking the below code that interferes with what the issue was geared towards. In the end this bit of code was from CodeDOM and I believe a lot of it was for readability.

File: src/Microsoft.AspNet.Razor/CodeGenerators/Visitors/CSharpCodeVisitor.cs [hunk 3]
```diff
@@ -16,6 +15,7 @@ public class CSharpCodeVisitor : CodeVisitor<CSharpCodeWriter>
         private const string ItemParameterName = "item";
         private const string ValueWriterName = "__razor_attribute_value_writer";
         private const string TemplateWriterName = "__razor_template_writer";
+        private const int MaxStringLiteralLength = 1024;
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/CodeGenerators/Visitors/CSharpCodeVisitor.cs]: 1024 to correspond with the http contexts response stream [buffer size](https://github.com/aspnet/Mvc/blob/47a3aee2c152b44bb75075e1f7494ba2124c82fa/src/Microsoft.AspNet.Mvc.Core/HttpResponseStreamWriter.cs#L24)

File: src/Microsoft.AspNet.Razor/CodeGenerators/CSharpCodeWriter.cs (hunk 2 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/CodeGenerators/CSharpCodeWriter.cs]: Suggest also removing `literal.Length >= 256 && literal.Length <= 1500` from [line 118](https://github.com/aspnet/Razor/blob/nimullen/stringconcats.614/src/Microsoft.AspNet.Razor/CodeGenerators/CSharpCodeWriter.cs#L118). A verbatim `string` is fine for anything lacking a `'\0'` character.

File: src/Microsoft.AspNet.Razor/CodeGenerators/Visitors/CSharpCodeVisitor.cs [hunk 4]
```diff
@@ -581,6 +561,31 @@ private CSharpCodeWriter RenderPreWriteStart()
...
+                charactersRendered += textToRender.Length;
+
+                if (Context.ExpressionRenderingMode == ExpressionRenderingMode.WriteToOutput)
+                {
+                    Writer.WriteEndMethodInvocation();
+                }
+            }
+        }
+
         public static CSharpCodeWriter RenderPreWriteStart(CSharpCodeWriter writer, CodeGeneratorContext context)
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/CodeGenerators/Visitors/CSharpCodeVisitor.cs]: Remove this `static` method and inline it into the unhelpful `RenderPreWriteStart()` method. Better if that's renamed to something meaningful e.g. `RenderStartWriteLiteralMethodInvocation()`. "pre write" is meaning-challenged.

File: src/Microsoft.AspNet.Razor/CodeGenerators/Visitors/CSharpCodeVisitor.cs [hunk 5]
```diff
@@ -581,6 +561,31 @@ private CSharpCodeWriter RenderPreWriteStart()
             return RenderPreWriteStart(Writer, Context);
         }
 
+        private void RenderChunkedWriteLiteral(string text)
+        {
+            var charactersRendered = 0;
+
+            // Render the string in pieces to avoid Roslyn OOM exceptions at compile time:
+            // https://github.com/dotnet/roslyn/issues/7398
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/CodeGenerators/Visitors/CSharpCodeVisitor.cs]: Use an indirection through aspnet/External. That's where we track non-aspnet issues and how they're referenced in our code.

File: src/Microsoft.AspNet.Razor/CodeGenerators/Visitors/CSharpCodeVisitor.cs [hunk 6]
```diff
@@ -581,6 +561,31 @@ private CSharpCodeWriter RenderPreWriteStart()
...
+            while (charactersRendered < text.Length)
+            {
+                if (Context.ExpressionRenderingMode == ExpressionRenderingMode.WriteToOutput)
+                {
+                    RenderPreWriteStart();
+                }
+
+                var charactersToSubstring = Math.Min(MaxStringLiteralLength, text.Length - charactersRendered);
+                var textToRender = text.Substring(charactersRendered, charactersToSubstring);
+                Writer.WriteStringLiteral(textToRender);
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/CodeGenerators/Visitors/CSharpCodeVisitor.cs]: Ick.
1. Don't `Substring()` if `text.Length <= MaxStringLiteralLength`.
2. Otherwise call `text.ToCharArray()` once before the loop and pass the array, start index, and length into the `Writer`.
3. Expose the necessary wrapper methods in `CodeWriter` (`WriteStringLiteral()`) and `CSharpCodeWriter` (`Write()`). Those APIs are anemic compared to the underlying `StringWriter` and lousy for this use case.

**Comment by dougbu** [comment]
:watch:

File: src/Microsoft.AspNet.Razor/CodeGenerators/Visitors/CSharpCodeVisitor.cs (hunk 6 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Razor/CodeGenerators/Visitors/CSharpCodeVisitor.cs]: I agree with the Doug. :maple_leaf:

**Comment by rynowak** [comment]
:watch:

File: src/Microsoft.AspNet.Razor/CodeGenerators/Visitors/CSharpCodeVisitor.cs (hunk 6 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/CodeGenerators/Visitors/CSharpCodeVisitor.cs]: Talked offline. Most of the perf changes for strings will be handled in [this](https://github.com/aspnet/Razor/issues/635) issue.

File: src/Microsoft.AspNet.Razor/CodeGenerators/CSharpCodeWriter.cs (hunk 2 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/CodeGenerators/CSharpCodeWriter.cs]: Chose to not remove the length restrictions to not muck with all the generated files in this PR. Also because the same code paths will be touched in [this](https://github.com/aspnet/Razor/issues/635) issue.

**Comment by NTaylorMullen** [comment]
:up::date:

File: src/Microsoft.AspNet.Razor/CodeGenerators/Visitors/CSharpCodeVisitor.cs [hunk 7]
```diff
@@ -576,25 +556,38 @@ private bool ShouldGenerateInstrumentationForExpressions()
...
+                    }
+                    else
+                    {
+                        Writer.WriteStartMethodInvocation(Context.Host.GeneratedClassContext.WriteLiteralMethodName);
+                    }
+                }
 
-            return writer;
+                var charactersToSubstring = Math.Min(MaxStringLiteralLength, text.Length - charactersRendered);
+                var textToRender = text.Substring(charactersRendered, charactersToSubstring);
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/CodeGenerators/Visitors/CSharpCodeVisitor.cs]: There's a difference between deferring perf work and adding a useless `Substring()` to the mainline code path. Keep the `text.Length < MaxStringLiteralLength` case. That's the code path that'll be used 99% of the time this method is called.

File: src/Microsoft.AspNet.Razor/CodeGenerators/CSharpCodeWriter.cs [hunk 8]
```diff
@@ -470,23 +470,6 @@ private void WriteCStyleStringLiteral(string literal)
...
-                        && (i < literal.Length - 1)
-                        && Char.IsLowSurrogate(literal[i + 1]))
-                    {
-                        Write(literal[++i].ToString());
-                    }
-
-                    Write("\" +");
-                    Write(NewLine);
-                    Write("\"");
-                }
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/CodeGenerators/CSharpCodeWriter.cs]: Note this change is unrelated to issue #614. Affects only `string`s (or the last `Substring`) of 256 `char`s or less, nothing that would hit the Roslyn issue.

Comment (dougbu) [inline src/Microsoft.AspNet.Razor/CodeGenerators/CSharpCodeWriter.cs]: Probably fine. Just worth mentioning separately in the squished (or squashed) commit description.

File: test/Microsoft.AspNet.Razor.Test/CodeGenerators/CSharpRazorChunkGeneratorTest.cs [hunk 9]
```diff
@@ -52,6 +52,7 @@ public void ConstructorAllowsEmptyRootNamespaceName()
         }
 
         [Theory]
+        [InlineData("StringLiterals")]
         [InlineData("NestedCSharp")]
         [InlineData("NullConditionalExpressions")]
         [InlineData("NestedCodeBlocks")]
```
Comment (dougbu) [inline test/Microsoft.AspNet.Razor.Test/CodeGenerators/CSharpRazorChunkGeneratorTest.cs]: Sort alphabetically :trollface:

File: test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/Output/StringLiterals.cs [hunk 10]
```diff
@@ -0,0 +1,238 @@
+#pragma checksum "StringLiterals.cshtml" "{ff1816ec-aa5e-4d10-87f7-6f4963833460}" "4bc2accb98c9c1f52bff719978da3f8449b7c9c6"
```
Comment (dougbu) [inline test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/Output/StringLiterals.cs]: This is one test of a brand new code path. Suggest adding a `CSharpCodeVisitorTest` or extending the `CSharpCodeGeneratorTest` class with a few unit tests at least for `Visit(ParentLiteralChunk)` and `Visit(LiteralChunk)`. Check near boundary conditions such as `string`s of length `0`, `1`, `255`, `256`, `257`, `1023`, `1024`, `1025`, ... Of course the middle few cases might be more interesting in the `CSharpCodeWriterTest` class.

File: src/Microsoft.AspNet.Razor/CodeGenerators/CSharpCodeWriter.cs (hunk 8 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/CodeGenerators/CSharpCodeWriter.cs]: Then again, no existing tests would change if you undid this. As-is, need a MVC reaction PR. Don't see one of those...

**Comment by dougbu** [comment]
:watch:
I don't have a solid recommendation in the [discussion of "extra" change](https://github.com/aspnet/Razor/pull/652#discussion-diff-49003626) but know a MVC reaction PR is needed as-is. Also looking for more tests and a fix [here](https://github.com/aspnet/Razor/pull/652#discussion-diff-49002871).

File: test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/Output/StringLiterals.cs (hunk 10 above)
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/Output/StringLiterals.cs]: This test covers most of those. Going to modify so 1 and 1023 are covered.

**Comment by NTaylorMullen** [comment]
:up: :date:

Also did not need to do an MVC reaction PR. Not many code gen tests and none of them have literals longer than 80 chars.

File: src/Microsoft.AspNet.Razor/CodeGenerators/Visitors/CSharpCodeVisitor.cs [hunk 11]
```diff
@@ -576,25 +556,47 @@ private bool ShouldGenerateInstrumentationForExpressions()
...
+                    }
+                    else
+                    {
+                        Writer.WriteStartMethodInvocation(Context.Host.GeneratedClassContext.WriteLiteralMethodName);
+                    }
+                }
 
-            return writer;
+                string textToRender;
+                if (text.Length <= MaxStringLiteralLength)
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/CodeGenerators/Visitors/CSharpCodeVisitor.cs]: Slightly odd to check this w/in a loop but :ok:

File: src/Microsoft.AspNet.Razor/CodeGenerators/CSharpCodeWriter.cs (hunk 2 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/CodeGenerators/CSharpCodeWriter.cs]: Guess you decided to stick w/ this change :8ball:

Comment (dougbu) [inline src/Microsoft.AspNet.Razor/CodeGenerators/CSharpCodeWriter.cs]: Oops, misplaced.

File: src/Microsoft.AspNet.Razor/CodeGenerators/CSharpCodeWriter.cs (hunk 8 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/CodeGenerators/CSharpCodeWriter.cs]: Guess you decided to stick w/ this change...

**Comment by dougbu** [comment]
:shipit:

File: src/Microsoft.AspNet.Razor/CodeGenerators/CSharpCodeWriter.cs (hunk 8 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/CodeGenerators/CSharpCodeWriter.cs]: Si :smile:

**Comment by rynowak** [comment]
:shipit:
